x-pack/plugins/ml/common/util/errors.ts

```diff
@@ -53,42 +53,118 @@ export interface MLCustomHttpResponseOptions<
   statusCode: number;
 }
 
-export const extractErrorMessage = (
+export interface MLErrorObject {
+  message: string;
+  fullErrorMessage?: string; // For use in a 'See full error' popover.
+  statusCode?: number;
+}
+
+export const extractErrorProperties = (
   error:
     | MLCustomHttpResponseOptions<MLResponseError | ResponseError | BoomResponse>
-    | undefined
     | string
-): string => {
-  // extract only the error message within the response error coming from Kibana, Elasticsearch, and our own ML messages
+    | undefined
+): MLErrorObject => {
+  // extract properties of the error object from within the response error
+  // coming from Kibana, Elasticsearch, and our own ML messages
+  let message = '';
+  let fullErrorMessage;
+  let statusCode;
 
   if (typeof error === 'string') {
-    return error;
+    return {
+      message: error,
+    };
+  }
+  if (error?.body === undefined) {
+    return {
+      message: '',
+    };
   }
-  if (error?.body === undefined) return '';
 
   if (typeof error.body === 'string') {
-    return error.body;
+    return {
+      message: error.body,
+    };
   }
   if (
     typeof error.body === 'object' &&
     'output' in error.body &&
     error.body.output.payload.message
   ) {
-    return error.body.output.payload.message;
+    return {
+      message: error.body.output.payload.message,
+    };
+  }
+
+  if (
+    typeof error.body === 'object' &&
+    'response' in error.body &&
+    typeof error.body.response === 'string'
+  ) {
+    const errorResponse = JSON.parse(error.body.response);
+    if ('error' in errorResponse && typeof errorResponse === 'object') {
+      const errorResponseError = errorResponse.error;
+      if ('reason' in errorResponseError) {
+        message = errorResponseError.reason;
+      }
+      if ('caused_by' in errorResponseError) {
+        const causedByMessage = JSON.stringify(errorResponseError.caused_by);
+        // Only add a fullErrorMessage if different to the message.
+        if (causedByMessage !== message) {
+          fullErrorMessage = causedByMessage;
+        }
+      }
+      return {
+        message,
+        fullErrorMessage,
+        statusCode: error.statusCode,
+      };
+    }
   }
 
   if (typeof error.body === 'object' && 'msg' in error.body && typeof error.body.msg === 'string') {
-    return error.body.msg;
+    return {
+      message: error.body.msg,
+    };
   }
 
   if (typeof error.body === 'object' && 'message' in error.body) {
+    if (
+      'attributes' in error.body &&
```

Comment:
is this check needed? if `attributes` is `undefined` the next `typeof` check will catch it.

Reply:
Yes this check is needed for type correctness.